In [35]:
import root_finding as rf
import math

In [36]:
#4
def func4(x):
    return x**3 - 3*x + 1

def dfunc4(x):
    return 3*x**2 - 3

print(rf.newton(func4, dfunc4, -0.8))

-0.8
1.8740740740740753
1.6140297543160267
1.5387263232855095
1.5321383803144402
1.5320888890233573
1.5320888862379562


Interpretation of the problem:
The tangent line at -0.8 led us to the right side of root 1.532. Therefore, the function just found root 1.532, rather than the root between 0 and 1. We can't find the desired root starting with -0.8.

In [37]:
#5
def func5(x):
    return x**3 - 2*x + 2

def dfunc5(x):
    return 3*x**2 - 2

print(rf.newton(func5, dfunc5, 0))

0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0


Interpretation of the problem:
The tangent line at 0 leads us to root x_1=1. The tangent line at 1 leads us back to 0. So x oscillates between 0 and 1 and never converges. We can't find the desired root starting with 0.

In [38]:
#6 starting from -3
def func6(x):
    return math.exp(x)/(1+math.exp(x)) - 0.5

def dfunc6(x):
    return math.exp(x)/((1 + math.exp(x))**2)

print(rf.newton(func6, dfunc6, -3))

-3
7.017874927409899
-551.187496891833


OverflowError: math range error

In [39]:
#6 starting from -2
print(rf.newton(func6, dfunc6, -2))

-2
1.626860407847019
-0.8188046331667729
0.09460983291123182
-0.00014120560728900222
4.693131191828293e-13


Interpretation of the problem:
The newton's method doesn't work when x_0=-3 but works when x_0=-2. The problem with -3 is the gradient at -3 is too flat so it leads us to a point with greater x (absolute) value, where the gradient is even flatter. And then we get to a point even farther away from the root and lastly there is a math range error because the number goes out of the range. 
-2 works because the gradient is steep enough for us to converge to 0. 

In [40]:
#7
def func7(x):
    return math.sin(x) - x

def dfunc7(x):
    return math.cos(x) - 1

print(rf.newton(func7, dfunc7, -2*math.pi))

Division by zero
None


Interpretation of the problem:
The Newton method works for most x values except integer multiples of 2π, where the slope becomes 0.

In [51]:
#8a
def func8a(x):
    return abs(x)**(1/3)

def dfunc8a(x):
    #print("evaluating derivative:", x)

    if x<0:
        # 1/3 * (-x)^(-2/3)
        return -((x**2)**(-1/3))/3
    if x>0:
        return (1/3)*(x**(-2/3))
    else:
        raise Exception("Attempted to evaluate derivative at 0")

print(rf.newton(func8a, dfunc8a, 1))

1
-2.0
4.0
-7.999999999999998
15.999999999999991
-31.99999999999998
63.99999999999992
-127.99999999999986
255.99999999999957
-511.999999999999
1023.9999999999974
-2047.9999999999936
4095.9999999999845
-8191.9999999999645
16383.99999999992
-32767.999999999825
65535.99999999959
-131071.99999999907
262143.9999999979
-524287.99999999534
1048575.9999999898
-2097151.9999999772
4194303.9999999492
-8388607.999999886
16777215.999999752
-33554431.999999456
67108863.9999988
-134217727.99999744
268435455.99999452
-536870911.9999882
1073741823.9999746
-2147483647.999946
4294967295.999885
-8589934591.999756
17179869183.999481
-34359738367.998894
68719476735.99766
-137438953471.99509
274877906943.98965
-549755813887.978
1099511627775.9534
-2199023255551.902
4398046511103.794
-8796093022207.565
17592186044415.082
-35184372088830.082
70368744177659.97
-140737488355319.6
281474976710638.47
-562949953421275.4
1125899906842547.4
-2251799813685088.5
4503599627370163.5
-9007199254740298.0
1.8014398509480542

Interpretation of the problem:
The gradient is too flat so the value diverges.

In [50]:
#8b
def func8b(x):
    return abs(x)**(2/3)

def dfunc8b(x):
    #print("evaluating derivative:", x)

    if x<0:
        return 2/3*(-x)**(-1/3)
    if x>0:
        return (2/3)*(x**(-1/3))
    else:
        raise Exception("Attempted to evaluate derivative at 0")
    
print(rf.newton(func8b, dfunc8b, 1))

1
-0.5
-1.25
-3.125
-7.8125
-19.53125
-48.82812499999999
-122.07031249999997
-305.1757812499999
-762.9394531249995
-1907.3486328124986
-4768.371582031245
-11920.92895507811
-29802.322387695265
-74505.80596923815
-186264.51492309532
-465661.2873077382
-1164153.2182693451
-2910383.0456733615
-7275957.6141834
-18189894.03545849
-45474735.0886462
-113686837.72161545
-284217094.30403847
-710542735.7600957
-1776356839.400238
-4440892098.500591
-11102230246.25147
-27755575615.62866
-69388939039.0716
-173472347597.67883
-433680868994.19666
-1084202172485.4907
-2710505431213.7246
-6776263578034.305
-16940658945085.746
-42351647362714.33
-105879118406785.72
-264697796016964.03
-661744490042409.4
-1654361225106021.8
-4135903062765050.0
-1.0339757656912612e+16
-2.5849394142281496e+16
-6.4623485355703656e+16
-1.6155871338925894e+17
-4.038967834731468e+17
-1.0097419586828657e+18
-2.524354896707161e+18
-6.310887241767894e+18
-1.5777218104419713e+19
-3.944304526104923e+19
-9.860761315262294e+19
-2.465

In [49]:
#8c
def func8c(x):
    return abs(x)**(4/3)

def dfunc8c(x):
    #print("evaluating derivative:", x)

    if x<0:
        return 4/3*(-x)**(1/3)
    if x>0:
        return (4/3)*(x**(1/3))
    else:
        raise Exception("Attempted to evaluate derivative at 0")
    
print(rf.newton(func8c, dfunc8c, 1))

1
0.25
0.0625
0.015624999999999993
0.003906249999999995
0.0009765624999999978
0.00024414062499999902
6.103515624999965e-05
1.5258789062499892e-05
3.8146972656249644e-06
9.536743164062394e-07
2.384185791015591e-07
5.960464477538962e-08
1.4901161193847365e-08


Interpretation of the problem:
Only the last one converges because the numerator is smaller than the denominator. The others don't converge at all no matter how close from 0 we start. 